x-pack/plugins/notifications/server/plugin.ts

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type {
+  PluginInitializerContext,
+  CoreSetup,
+  CoreStart,
+  Plugin,
+  Logger,
+} from '@kbn/core/server';
+import type {
+  NotificationsPluginSetupDeps,
+  NotificationsPluginStartDeps,
+  NotificationsPluginSetup,
+  NotificationsPluginStart,
+} from './types';
+import { EmailService } from './services';
+import { NotificationsConfigType } from './config';
+import { getResolvablePromise, type ResolvablePromise } from '../common/lib';
+import { PLUGIN_ID } from '../common';
+
+export class NotificationsPlugin
+  implements Plugin<NotificationsPluginSetup, NotificationsPluginStart>
+{
+  private readonly logger: Logger;
+  private readonly initialConfig: NotificationsConfigType;
+  private email: ResolvablePromise<EmailService>;
+  private emailConnector: string;
+
+  constructor(initializerContext: PluginInitializerContext<NotificationsConfigType>) {
+    this.logger = initializerContext.logger.get();
+    this.email = getResolvablePromise();
+    this.initialConfig = initializerContext.config.get();
+  }
+
+  public setup(core: CoreSetup, plugins: NotificationsPluginSetupDeps) {
+    // TODO this must be defaulted to 'elastic-cloud-email' for cloud
+    if (!plugins.actions) {
+      this.email.doReject(`Error starting notification services: 'actions' plugin not available.`);
+      return { email: this.email };
```

Comment:
nit: Early returns is a clean and nice pattern:

```ts
if (something) {
  return early;
}

return late;
```

But here it is a bit risky to use, especially as plugin classes are usually not tested. The reason being that once this plugin has more services, it will not be possible to use early returns in this manner and it will need to be refactored:


```ts
setup() {
  if (!plugins.actions) {
    return { email: this.email }; // BUG: "sms" service is missing.
  }

  // ... more code ...

  return {
    email: this.email,
    sms: this.sms,
  };
}
```